Added flatpak

Jakkumn opened the pull request:
Hi all,
I've been running an immutable OS and wanted to see if I could figure out a flatpak install solution for this. 
Key commands:
# Build for testing
flatpak-builder --force-clean build-dir com.timothycrosely.streamdeck-ui.json
# Run testing
flatpak-builder --run build-dir com.timothycrosely.streamdeck-ui.json streamdeck
# Build for deployment
flatpak-builder --repo=repo --force-clean build-dir com.timothycrosely.streamdeck-ui.json
#  Add personal repo that was pushed to
flatpak --user remote-add --no-gpg-verify personal repo
# Install from personal repo
flatpak --user install personal com.timothycrosley.streamdeck-ui
# Run installed flatpak
flatpak run com.timothycrosley.streamdeck-ui

It added amixer and the ability to connect to mpris media player. There are probably other items that need to be installed to connect to other entities, but this fit my minimal use case of managing volume and the media player.

4s3ti commented:
Hi, thanks a lot for your PR, this can come in quite handy for a lot of ppl and has been highly requested. 

please change the naming from com.timothycrosley.streamdeck-ui to com.streamdeck-linux-gui, or some other naming that feels more algnined, sice @timothycrosley seems to have vanished and we had to fork it makes very little sense to keep application related stuff with a personal name on it. 

I Will test this as soon as I have a bit. most likely during the weekend.

joaojotta commented:
Please add this. I've been testing different systems on a laptop to see if I can set up a "studio OS" when I need to use my Stream Deck and I can't even build it following the commands. I keep getting a "pip failed to build package: evdev" error.
Maybe a flatpak would allow me to run the WONDERFULL app. 😃

4s3ti commented:
hi, sorry for the delay ... but been busy IRL ... have you tested it on Xorg on Wayland? 

It does not run for me on Wayland/Sway. 


```
$flatpak-builder --run build-dir build-aux/com.streamdeck.linux-gui.json streamdeck
qt.qpa.xcb: could not connect to display
qt.qpa.plugin: From 6.5.0, xcb-cursor0 or libxcb-cursor0 is needed to load the Qt xcb platform plugin.
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vkkhrdisplay, vnc, wayland-egl, wayland, xcb.
```

Also noticed instructions are not super accurate. 

To build the package, necessary dependencies are required :

org.freedesktop.Platform 
org.freedesktop.Sdk

Also .. a change to the name is required, it fails to build as it requires 2 periods.
I suggest to chage it to:

com.streamdeck.linux-gui 

Also, your instructions assume whoever is building it is inside the build-aux dir.  please update the instructions to be `... build-dir build-aux/com.streamdeck.linux-gui.json`


Once again really appreciated and really sorry for taking so long to look at this.

Jakkumn commented:
@4s3ti 
No worries, thank you for reviewing and telling me whats up. Sorry for the thrash on the naming... I haven't built flatpaks before.

I had changed from freedesktop to kde to install qt (which should resolve your output snippet above - xcb-cursor provides a link to qt as far as I can tell). 

I also added a readme to the build-aux folder. Left a note about prepending other pathing as needed. Please give it another shot and let me know if I have missed anything else. Happy to fix and get it running.

Hope to hear back soon.

4s3ti commented on build-aux/com.foursthreeti.streamdeck-linux-gui.json:
Some weird copy pasta? :D 

Actually instead of my previous suggestion should go for this naming: 

`com.streamdeck-linux.gui`

Jakkumn replied:
No, weird requirements for the first pathing. :P Can't start w/ the number 4. Also can't have dashes in the first part. Renamed to: com.streamdeck.linux-gui.json
Re-running all instructions on my end again... give me a few.

Jakkumn replied:
Unless there is a different name we should use... figured shouldn't use 'com.elgato.xxxxx' since this project isn't them and I don't want to seem like it is supported by them.

4s3ti commented on build-aux/com.foursthreeti.streamdeck-linux-gui.json:
```diff
@@ -0,0 +1,127 @@
+{
+    "app-id": "com.foursthreeti.streamdeck-linux-gui",
```
Same as before here

4s3ti reviewed (commented):
Pleas check this ones, let me know if there's any issue with this naming. 

let's avoid linking any naming to individuals. 


I will test the rest in the meantime

4s3ti commented:
> @4s3ti
> No worries, thank you for reviewing and telling me whats up. Sorry for the thrash on the naming... I haven't built flatpaks before.
> 
> I had changed from freedesktop to kde to install qt (which should resolve your output snippet above - xcb-cursor provides a link to qt as far as I can tell).
>
>I also added a readme to the build-aux folder. Left a note about prepending other pathing as needed. Please give it another shot and let me know if I have missed anything else. Happy to fix and get it running.
>
>New readme: https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/blob/311f8df805fd2fe81301fa753cb0399dc44463cc/build-aux/README.md
>
>Hope to hear back soon.


No worries, you've managed to do get much further than me or anyone else around here ;) doing just fine, thank you for your effort.

Jakkumn commented:
Latest is naming changes + a note in the readme for how to uninstall. If you can install, its nice to be able to uninstall. :P

4s3ti commented:
Still no joy :( 

```
$flatpak-builder --run build-dir build-aux/com.streamdeck.linux-gui.json streamdeck
qt.qpa.xcb: could not connect to display
qt.qpa.plugin: From 6.5.0, xcb-cursor0 or libxcb-cursor0 is needed to load the Qt xcb platform plugin.
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vkkhrdisplay, vnc, wayland-egl, wayland, xcb.
``` 


I tried changing the build file and add  the `QT_QPA_PLATFORM` variable, but still get the same issue ... with just the different plugin name in the error:

```
    "finish-args": [
        "--share=ipc",
        "--env=QT_QPA_PLATFORM=wayland",
        "--socket=fallback-x11",
        "--socket=pulseaudio",
        "--env=PULSE_PROP_media.category=Manager",
        "--device=all",
        "--talk-name=org.mpris.MediaPlayer2.*",
        "--filesystem=home"
    ],
```


```
$flatpak-builder --run build-dir build-aux/com.streamdeck.linux-gui.json streamdeck
Failed to create wl_display (No such file or directory)
Fontconfig error: Cannot load config file "/run/host/font-dirs.xml": No such file: /run/host/font-dirs.xml
qt.qpa.plugin: Could not load the Qt platform plugin "wayland" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vkkhrdisplay, vnc, wayland-egl, wayland, xcb.
```

... there's something more missing in the build to support native wayland sessions. For context I don't have XonWayland and I refuse to, there should be no reason for anyone to still use any form of XSession :)

Jakkumn commented:
I may be a dummy here, I think I forgot to add the '--socket=wayland' to finish args. Just a sec

Jakkumn commented:
Alrighty, I think I have it alive. With your note on the env being set to wayland (instead of defaulting to xcb), adding the wayland socket (turns out fallback x11 doesn't allow wayland through...), and maybe? updating the flatpak version of qt to 6.6 and it seemed to work. Give it another shot!

Jakkumn commented:
Only note, it may cause the x11 fallback to break... I will keep investigating how to get both paths to work.

4s3ti commented:
Making some progress, GJ!!! 

Application now launches, however interacting with system applications via command is not working .. for example: 

```
The command 'wshowkeys -a top -a right -F 'Cantarell 24'' failed: [Errno 2] No such file or directory: 'wshowkeys'
The command 'killall wshowkeys' failed: [Errno 2] No such file or directory: 'killall'
The command 'wl-copy 'foo'' failed: [Errno 2] No such file or directory: 'wl-copy'
The command 'pavucontrol' failed: [Errno 2] No such file or directory: 'pavucontrol'
```

* Executing scripts in the home dir works, 
* Providing the full path to the Binaries, for example `/usr/bin/wshowkeys` doesn't work either. 
* Tray Icon does not work

Looking at the requirements about flathub submission ... there's one thing that I am not sure if its done on the flatpak side or should be done on the pip package side: https://docs.flathub.org/docs/for-app-authors/requirements#desktop-integration

4s3ti commented:
hey @Jakkumn  did you had any time to look into this? Thank you :)

coolapso commented:
https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/discussions/203
